### PR TITLE
Phi-architechture + other features

### PR DESCRIPTION
**Pull request by IsNoobgrammer**
Added support for Phi architechture

Added Wandb logging

Sample:- 
image

Enhanced Support for LoRA
Adafactor,Lion Optimizers added , adam is now optimized
Optimizers now fully support Weight-decay
Added cosine scheduling , enhanced linear-scheduling 
Schedulars now fully support Warmup-ratio

### CODE REVIEW

**Comment by Locutusque** [comment]
Thank you for these contributions! This looks safe to pull.

**Comment by IsNoobgrammer** [comment]
> Thank you for these contributions! This looks safe to pull.

I will try to update it more features and lets hope I don't make any typos 
Currently on TODO are 

-Support DPO
-Support Gradient Accumulation
-Support Quantization
-Support for 8-bit optimizers and paged optimizers
-Support interference after sharding
-Implement Bleu & Perpexility (DPO)

**Comment by Locutusque** [comment]
I was also thinking I should implement an evaluation script for LLMs.

**Comment by IsNoobgrammer** [comment]
> I was also thinking I should implement an evaluation script for LLMs.

Yes it would be so nice to have native "llm-leaderboard"   (types)
would be much easier then to see if model converged properly or was overfitted 
interference seems to be an entity of distant future for now

**Comment by IsNoobgrammer** [comment]
> It might be a good look into this repo for DPO: https://github.com/eric-mitchell/direct-preference-optimization. Not sure how we're going to implement quantization for TPU though.

Since bnb is not supported on CPU/TPU
but there is this new quantization ,method [AQLM](https://huggingface.co/docs/transformers/main/en/quantization#aqlm) 
it looks promising as it supports quantization on cpu
